tests/unit/instance-initializers/prefetch-test.js

```diff
@@ -48,3 +49,79 @@ test('the prefetch hook is invoked', function(assert) {
   assert.equal(handler1._prefetched, promise, 'the promise is set on handler1 as _prefetched');
   assert.equal(handler2._prefetched, promise, 'the promise is set on handler2 as _prefetched');
 });
+
+test('the prefetch hook is invoked for async handlers', function(assert) {
+  assert.expect(4);
+
+  initialize(instance);
+
+  const router = instance.lookup('router:main');
+  const promise = new Ember.RSVP.resolve(1);
+  const runSharedModelHook = this.stub().returns(promise);
+  const handler1 = {};
+  const handler2 = {};
+  const handlerInfos = [{
+    runSharedModelHook,
+    handler: handler1,
+  }, {
+    runSharedModelHook,
+    handler: undefined,
+    handlerPromise: new Ember.RSVP.resolve(handler2)
+  }];
+  const transition = { handlerInfos };
+
+  router.trigger('willTransition', transition);
+
+  return wait().then(() => {
+    assert.ok(runSharedModelHook.calledTwice, 'handlerInfo.runSharedModelHook is called once per handlerInfo');
+    assert.ok(runSharedModelHook.calledWith(transition, 'prefetch'), 'handlerInfo.runSharedModelHook is called with the transition and "prefetch"');
+    assert.equal(handler1._prefetched, promise, 'the promise is set on handler1 as _prefetched');
+    assert.equal(handler2._prefetched, promise, 'the promise is set on handler2 as _prefetched');
+  });
+});
+
+test('the prefetch hook is invoked according to route hierarchy for async handlers', function(assert) {
+  assert.expect(1);
+
+  initialize(instance);
+
+  const router = instance.lookup('router:main');
+  const operations = [];
+  const handler1 = {};
+  const handler2 = {};
+  const handlerInfos = [{
+    runSharedModelHook() {
+      operations.push('prefetch 1');
+    },
+    handler: undefined,
+    handlerPromise: new Ember.RSVP.Promise((resolve) => {
+      setTimeout(() => {
+        operations.push('resolved handler 1');
+        resolve(handler1);
+      }, 50);
+    })
+  }, {
+    runSharedModelHook() {
+      operations.push('prefetch 2');
+    },
+    handler: undefined,
+    handlerPromise: new Ember.RSVP.Promise((resolve) => {
+      setTimeout(() => {
+        operations.push('resolved handler 2');
+        resolve(handler2);
+      }, 0);
+    })
+  }];
+  const transition = { handlerInfos };
+
+  router.trigger('willTransition', transition);
+
+  return transition._handlerPromiseChain.then(() => {
```

Comment:
I see the issue. What about scheduling a MacroTask to perform the comparison?

```javascript
return new Ember.RSVP.Promise((resolve) => {
  setTimeout(() => {
    assert.deepEqual(operations, [
      'resolved handler 2',
      'resolved handler 1',
      'prefetch 1',
      'prefetch 2'
    ]);
    resolve();
  }, 0);
});
```

I've pulled down your code and tested the above, and it works. Thoughts? I'm thinking it's better for the test to be ugly than to muck with the transition object, and we can still dodge the 50ms test slowdown. Could optionally create a wrapper to make the intention more explicit:

```javascript
function macroTask(callback) {
  setTimeout(callback, 0);
}
```